### PR TITLE
Parallelize text search

### LINKED ISSUE #15384: File Search - Improve performance of file search
Part of #55.

We will focus on simple text search. Regex text searches might be more difficult to optimize and are used less often than simple text searches. We will initially explore our options (e.g., refining our own implementation, command line tools, text search libraries/engines).

We will include full-text search libraries/engines in our investigation and see if any of these can do simple text searches while at the same time open the door for also supporting a more natural 'fuzzy' text search at a later time.

The goal is raw search performance with library size (if any) and disk usage of an index (if any) being potential trade-offs.

### PR DESCRIPTION
**Pull request by roblourens**
Fixes https://github.com/Microsoft/vscode/issues/15384#issuecomment-262605055, comment is updated with final timings.

This PR rearranges the existing text search code to run using multiple processes. This doesn't include work to show results sooner - users still need to wait for a batch of 512 file matches to be filled. That work is tracked in Microsoft/vscode#16284. Here's a brain dump for posterity -

RawSearchService creates an instance of textSearch.Engine and passes it a FileWalker which is producing paths in the workspace. This is the same.

textSearch.Engine creates some number of Search Worker processes. The number is determined by `require('os').cpus().length` which returns 8 on a MBP, because it has 4 physical cores, with hyperthreading.

The fileWalker produces file paths, and the sizes in bytes of those files. They are grouped into batches of roughly equal sizes in bytes. The number of files in each batch may vary a lot, but each batch should represent files that are cumulatively about 1MB in size. Hopefully this will make them more likely to be processed in a similar amount of time, with a similar number of results, so one process isn't running for much longer than others. This might sound like a large batch - smaller workspaces won't benefit much from parallelization - but they don't need it. I did some testing on a few workspaces of different sizes with different batch sizes, and 1MB seemed like the sweet spot. I also experimented with a batch size that starts small and grows exponentially, thinking it might help on medium-sized workspaces, but it didn't seem to make a difference.

Batches are sent to the worker as soon as they're ready, in a round-robin fashion. I experimented with sending one batch at a time to each process, with another when it's finished, but that was slower because the worker isn't doing anything while it's waiting for a new batch. I also tried sending a couple batches up front, and more as needed, but it was no better than sending them ASAP. Looking at timings, I think this is because the workers are slow to handle the first couple batches they receive - the fileWalker is actually often done by the time the workers are sending back their first batch of results. I need to investigate more to figure out why this is - maybe because the new processes are still loading code, JITting, etc. They process the rest of the batches much faster after those first few. It may be faster to keep the worker processes around between searches, but I don't know that we want to persist 8 (or more) processes per vscode window.

The worker searches the files using the same code that we already had, this hasn't changed much. It does include the change to eliminate Buffer.slice calls that is already checked in. The only change is that I limited the number of files that it will open at a time, because I kept running into the OS open file limit without it.

The batch includes the current `maxResults`, i.e. the initial max minus the number of results received so far. The main process might receive more than the max number of results from workers, so it has to trim the result set. This is kind of annoying to get exactly right, because the result set is not a flat list, but a tree of files with matches, lines in those files with matches, and the matches on each line. So, it will just take batches until the total is greater than the max. It can go over the max, but this is much easier.

### CODE REVIEW

File: src/vs/workbench/services/search/node/worker/searchWorkerIpc.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	call(command: string, arg?: any): TPromise<any>;
+}
+
+export class SearchWorkerChannel implements ISearchWorkerChannel {
+	constructor(private worker: SearchWorker) {
+	}
+
+	call(command: string, arg?: any): TPromise<any> {
+		switch(command) {
+			case 'initialize': return TPromise.wrap(this.worker.initialize(arg));
```
Comment (chrmarti) [inline src/vs/workbench/services/search/node/worker/searchWorkerIpc.ts]: `TPromise.wrap` is already used in initialize().

File: src/vs/workbench/test/electron-browser/textsearch.perf.test.ts [hunk 2]
```diff
@@ -114,7 +118,7 @@ suite('TextSearch performance', () => {
 		const finishedEvents = [];
 		return runSearch() // Warm-up first
 			.then(() => {
-				if (testWorkspaceArg) { // Don't measure by default
+				if (testWorkspacePath) { // Don't measure by default
```
Comment (chrmarti) [inline src/vs/workbench/test/electron-browser/textsearch.perf.test.ts]: Btw., any unit tests that would make sense adding or are the existing ones still sufficient?

File: src/vs/workbench/services/search/node/textSearch.ts [hunk 3]
```diff
@@ -5,330 +5,178 @@
...
+			this.processedBytes += processed;
 
 			// Emit progress() unless we got canceled or hit the limit
 			if (processed && !this.isDone && !this.isCanceled && !this.limitReached) {
 				progress();
 			}
 
 			// Emit done()
-			if (this.worked === this.total && this.walkerIsDone && !this.isDone) {
+			if (!this.isDone && this.processedBytes === this.totalBytes && this.walkerIsDone) {
```
Comment (roblourens) [inline src/vs/workbench/services/search/node/textSearch.ts]: No, we only check the file size once. This is based on the size that the fileWalker provides. I could change it to >= though.

File: src/vs/workbench/services/search/node/textSearch.ts [hunk 4]
```diff
@@ -5,330 +5,178 @@
...
-	private readlinesAsync(filename: string, perLineCallback: (line: string, lineNumber: number) => void, options: ReadLinesOptions, callback: (error: Error) => void): void {
-		fs.open(filename, 'r', null, (error: Error, fd: number) => {
-			if (error) {
-				return callback(error);
+		// Walk over the file system
+		let nextBatch = [];
+		let nextBatchBytes = 0;
+		const batchFlushBytes = 2 ** 20; // 1MB
+		this.walker.walk(this.config.rootFolders, this.config.extraFiles, result => {
+			let bytes = result.size || 1;
```
Comment (roblourens) [inline src/vs/workbench/services/search/node/textSearch.ts]: Maybe so we continue to report progress while searching empty files. It won't break that check though.

File: src/vs/workbench/services/search/node/textSearch.ts [hunk 5]
```diff
@@ -5,330 +5,178 @@
...
-
-				return unwind(size);
-			};
-
-			const absolutePath = result.base ? [result.base, result.relativePath].join(path.sep) : result.relativePath;
-			let perLineCallback = (line: string, lineNumber: number) => {
-				if (this.limitReached || this.isCanceled) {
-					return; // return early if canceled or limit reached
+			const maxResults = this.config.maxResults - this.numResults;
+			worker.search({ absolutePaths: batch, maxResults }).then(result => {
```
Comment (roblourens) [inline src/vs/workbench/services/search/node/textSearch.ts]: Previously we didn't surface errors from the text search, just from the file walker. I wasn't sure whether to change that. I guess I should do something with the error.

File: src/vs/workbench/services/search/node/textSearch.ts (hunk 3 above)
Comment (chrmarti) [inline src/vs/workbench/services/search/node/textSearch.ts]: Got it. Should be fine as-is then.

File: src/vs/workbench/test/electron-browser/textsearch.perf.test.ts (hunk 2 above)
Comment (chrmarti) [inline src/vs/workbench/test/electron-browser/textsearch.perf.test.ts]: We have some unit tests in search.test.ts.

**Comment by bpasero** [comment]
Unassigning myself, I have full trust in @chrmarti review. If there are specific questions, ping me.

**Comment by roblourens** [comment]
Thanks, merging

**Comment by sallar** [comment]
Hi @roblourens, just wanted to say that this code reads like poetry to me. Thank you.

**Comment by roblourens** [comment]
Thanks @sallar :)

**Pull request merged**